jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint/ConstraintComponentSPARQL.java

```diff
@@ -59,6 +59,18 @@ public ConstraintComponentSPARQL(SparqlComponent sparqlConstraintComponent,
         }
     }
 
+    public SparqlComponent getSparqlConstraintComponent() {
+        return sparqlConstraintComponent;
+    }
+
+    public Multimap<Parameter, Node> getParameterMap() {
+        return parameterMap;
```

Comment:
Not sure if there's any risk here (or in the other getters that return objects and lists) to have concurrency issues. If so, we could copy the underlying object/collection.

Reply:
Yes, it does matter. `.equals` relies on all parts of the shape and shapes go into datastructures so `hashCode` and `equals` must be stable. A `Collections.unmodifiable` (on "get" or in the constructor, not sure which is better) or a structure copy is needed. The current code probably isn't perfect.

Or documentation!

Another to look at is whether internal calculated members should be made public.

Reply:
Sorry, good catch. I think wrapping with unmodifiable collections or cloning in the getter should do. I'll add another commit.

Reply:
For the `Multimap` in `ConstraintComponentSPARQL` I am not sure how to proceed, I think it's best to revert the change on that one.

Reply:
>For the Multimap in ConstraintComponentSPARQL I am not sure how to proceed, I think it's best to revert the change on that one.

If that's OK for you for now, then all good :-)